lambda/matching.ml

```diff
@@ -3476,17 +3484,9 @@ and compile_match_nonempty ~scopes repr partial ctx
   | _ -> assert false
 
 and compile_match_simplified ~scopes repr partial ctx
-    (m : Simple.clause pattern_matching) =
-  match m with
-  | { cases = []; args = [] } -> comp_exit ctx m
-  | { args = ((Lvar v as arg), str) :: argl } ->
-      bind_match_arg str v arg (
-        let args = (arg, Alias) :: argl in
-        let m = { m with args } in
-        let first_match, rem = split_and_precompile_simplified m in
-        combine_handlers ~scopes repr partial ctx first_match rem
-      )
-  | _ -> assert false
```

Comment:
I'm not sure I understand why this PR allows this simplification or makes it more obviously correct. Does the new `simple_arg` shape tell us something more specific about `m.args`? Or was this simplification possible before?

(Separately, I'll note that `bind_match_arg` is now called only once, and so could be inlined if you found that more readable.)

Reply:
Before we would need a dynamic check to ensure that the list of arguments is not empty, and we know this statically now. The code was also written as if the argument needed to be bound, while we can now statically reason about the fact that it is pure and its `let_kind` does not matter.

One aspect of the change that is subtle is that we now allow the `Tuple` case to reach here, but in fact I believe that this case cannot happen. (So we could add an `assert false` if we wanted to, as we do here.) I think that the Tuple case can only flow through the precompilation logic, never to this actual compilation code due to the different `compile_flattened` code path. I tried to check this statically, but the cost in complexity is quite high (polymorphic variants, polymorphic recursion, oh my!) and I did not finish it. I prefer the new version which might implicitly generate less-good-than-we-expected code if a `Tuple` argument somehow found its way here.

Reply:
This seems fine then.

> Before we would need a dynamic check to ensure that the list of arguments is not empty, and we know this statically now.
Agreed, this is clear.

> The code was also written as if the argument needed to be bound, while we can now statically reason about the fact that it is pure and its let_kind does not matter.

It seems to me this was already true as of this refactor: https://github.com/ocaml/ocaml/commit/5bffef52815539e0f7d071f780427a2dfe8174b1. The call to `bind_match_arg` is guarded by a match on `Lvar v`. I'm saying this just to check my understanding, not to request changes.

Reply:
Yes, I *think* that I could have simplified the `compile_match_simplified` part of the change by removing the `bind` logic at that point. I feel more comfortable with the change now that the types make it clear that the argument list has already been "split" by the time we reach this point in the code -- this is a sharp distinction with `compile_match_nonempty` above.